src/cargo/util/semver_ext.rs

```diff
@@ -171,12 +171,12 @@ impl std::str::FromStr for PartialVersion {
         let version_req = match semver::VersionReq::parse(value) {
             // Exclude semver operators like `^` and pre-release identifiers
             Ok(req) if value.chars().all(|c| c.is_ascii_digit() || c == '.') => req,
-            _ if value.contains('+') => {
-                anyhow::bail!("unexpected build field, expected a version like \"1.32\"")
-            }
             _ if value.contains('-') => {
                 anyhow::bail!("unexpected prerelease field, expected a version like \"1.32\"")
             }
+            _ if value.contains('+') => {
+                anyhow::bail!("unexpected build field, expected a version like \"1.32\"")
+            }
```

Comment:
Do we need to check for this? `VersionReq` cannot parse requirements that have build metadata? And why are we parsing the string to determine what kind of requirement we got and not looking at the `req` we matched on?